src/callbacks/save_solution.jl

```diff
@@ -0,0 +1,83 @@
+
+# TODO: Taal, implement, save AMR indicator values
+# TODO: Taal, refactor, allow saving arbitrary functions of the conservative variables
+mutable struct SaveSolutionCallback
+  save_initial_solution::Bool
+  save_final_solution::Bool
+  output_directory::String
+  solution_variables::Symbol
+end
+
+
+function Base.show(io::IO, cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:SaveSolutionCallback}
+  stepsize_callback = cb.affect!
+  print(io, "SaveSolutionCallback")
+end
+# TODO: Taal bikeshedding, implement a method with more information and the signature
+# function Base.show(io::IO, ::MIME"text/plain", cb::DiscreteCallback{Condition,Affect!}) where {Condition, Affect!<:StepsizeCallback}
+# end
+
+
+function SaveSolutionCallback(; solution_interval=0,
+                                save_initial_solution=true,
+                                save_final_solution=true,
+                                output_directory="out",
+                                solution_variables=:primitive)
+  condition = (u, t, integrator) -> solution_interval > 0 && ((integrator.iter % solution_interval == 0) || (save_final_solution && t == integrator.sol.prob.tspan[2]))
+
+  solution_callback = SaveSolutionCallback(save_initial_solution, save_final_solution,
+                                           output_directory, solution_variables)
+
+  DiscreteCallback(condition, solution_callback,
+                   save_positions=(false,false),
+                   initialize=initialize!)
+end
+
+
+function initialize!(cb::DiscreteCallback{Condition,Affect!}, u, t, integrator) where {Condition, Affect!<:SaveSolutionCallback}
+  reset_timer!(timer())
+  solution_callback = cb.affect!
+
+  mkpath(solution_callback.output_directory)
+
+  semi = integrator.p
+  mesh, _, _, _ = mesh_equations_solver_cache(semi)
+  mesh.unsaved_changes = true
+
+  if solution_callback.save_initial_solution
+    solution_callback(integrator)
+  end
+
+  return nothing
+end
+
+
+function (solution_callback::SaveSolutionCallback)(integrator)
+  @unpack u, t, dt, iter = integrator
```

Comment:
Is `integrator` a Trixi type or a DiffEq type?

Reply:
DiffEq type